fish-rust/src/wildcard.rs

```diff
@@ -11,3 +33,1259 @@ pub const ANY_STRING_RECURSIVE: char = char_offset(WILDCARD_RESERVED_BASE, 2);
+                    assert!(next_slash.is_some());
+                    let mut prefix = effective_prefix.to_owned() + wc_segment;
+                    prefix.push('/');
+                    self.expand_intermediate_segment(
+                        base_dir,
+                        &mut dir,
+                        wc_segment,
+                        wc_remainder,
+                        &prefix,
+                    );
+                }
+
+                let Some(asr_idx) = wc_segment.find_char(ANY_STRING_RECURSIVE) else {
+                    return;
+                };
+
+                // Apply the recursive **.
+                // Construct a "head + any" wildcard for matching stuff in this directory, and an
+                // "any + tail" wildcard for matching stuff in subdirectories. Note that the
+                // ANY_STRING_RECURSIVE character is present in both the head and the tail.
+                let head_any = wc_segment.slice_to(asr_idx + 1);
+                let any_tail = wc.slice_from(asr_idx);
+                assert!(head_any.chars().last().unwrap() == ANY_STRING_RECURSIVE);
+                assert!(any_tail.char_at(0) == ANY_STRING_RECURSIVE);
+
+                dir.rewind();
+                self.expand_intermediate_segment(
+                    base_dir,
+                    &mut dir,
+                    head_any,
+                    any_tail,
+                    effective_prefix,
+                );
+            }
+        }
+
+        pub fn status_code(&self) -> WildcardResult {
+            if self.did_interrupt {
+                return WildcardResult::Cancel;
+            } else if self.did_overflow {
+                return WildcardResult::Overflow;
+            }
+            match self.did_add {
+                true => WildcardResult::Match,
+                false => WildcardResult::NoMatch,
+            }
+        }
+    }
+
+    impl<'receiver, 'c> WildCardExpander<'receiver, 'c> {
+        /// We are a trailing slash - expand at the end.
+        fn expand_trailing_slash(&mut self, base_dir: &wstr, prefix: &wstr) {
+            if self.interrupted_or_overflowed() {
+                return;
+            }
+
+            if !self.flags.contains(ExpandFlags::FOR_COMPLETIONS) {
+                // Trailing slash and not accepting incomplete, e.g. `echo /xyz/`. Insert this file, we already know it exists!
+                self.add_expansion_result(base_dir.to_owned());
+                return;
+            }
+            // Trailing slashes and accepting incomplete, e.g. `echo /xyz/<tab>`. Everything is added.
+            let Ok(mut dir) = self.open_dir(base_dir, false) else {
+                return;
+            };
+
+            // wreaddir_resolving without the out argument is just wreaddir.
+            // So we can use the information in case we need it.
+            let need_dir = self.flags.contains(ExpandFlags::DIRECTORIES_ONLY);
+
+            while let Some(Ok(entry)) = dir.next() {
```

Comment:
This had an unused `path` 
https://github.com/fish-shell/fish-shell/blob/ed881bcdd8d7351c5b48daf22ca6e3d9344c2d5d/src/wildcard.cpp#L692C1-L692C1

Introduced by this:

https://github.com/fish-shell/fish-shell/commit/a2d816710fc336e0a964e11b798cd5fb19043760#diff-d6f14992d417e6d5eb82c1590cd41b90fa804b60fd38ad6c2cf9efb66d73d1f5L740